test/distributed/org/apache/cassandra/distributed/test/ReadFailureTest.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.test;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.junit.Test;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.distributed.api.ICluster;
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
+import org.apache.cassandra.exceptions.RequestFailureReason;
+
+public class ReadFailureTest extends TestBaseImpl
```

Comment:
When I run this test locally (without limiting resources) I get

```
$ ./ci-test org/apache/cassandra/distributed/test/ReadFailureTest
...
testclasslist:
     [echo] Number of test runners: 1
    [mkdir] Created dir: /Users/dcapwell/src/github/apache/cassandra/team/aholmberg/build/test/cassandra
    [mkdir] Created dir: /Users/dcapwell/src/github/apache/cassandra/team/aholmberg/build/test/output
[junit-timeout] Picked up _JAVA_OPTIONS: -Djava.net.preferIPv4Stack=true
[junit-timeout] Testsuite: org.apache.cassandra.distributed.test.ReadFailureTest
[junit-timeout]
[junit-timeout] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "ReadStage-1"
[junit-timeout]
[junit-timeout] Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "ReadStage-2"
[junit-timeout] Testsuite: org.apache.cassandra.distributed.test.ReadFailureTest Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 99.03 sec
[junit-timeout]
[junit-timeout] Testcase: testSpecExecRace(org.apache.cassandra.distributed.test.ReadFailureTest):      FAILED
[junit-timeout] null
[junit-timeout] junit.framework.AssertionFailedError
[junit-timeout]         at org.apache.cassandra.distributed.test.ReadFailureTest.loopFailStatement(ReadFailureTest.java:95)
[junit-timeout]         at org.apache.cassandra.distributed.test.ReadFailureTest.testSpecExecRace(ReadFailureTest.java:76)
[junit-timeout]
[junit-timeout]
[junit-timeout] Test org.apache.cassandra.distributed.test.ReadFailureTest FAILED

BUILD FAILED
```

Looks like the test is flaky, can you take a look?

Reply:
I had not seen that, but will definitely take a look. Thanks.

Reply:
Looking at the instance logs I see

```
ERROR [node1_isolatedExecutor:18] node1 2020-12-03 12:37:35,969 OutOfMemory error letting the JVM handle the error:
java.lang.OutOfMemoryError: Java heap space
```

So looks like we ran out of direct memory, which then leads to JVMStabilityInspector trying to convert to heap (to trigger JVM crash); so we are actually running out of direct memory it seems.

Reply:
patched the build to not hid stack traces, I see the following

```
ERROR [ReadStage-2] node1 2020-12-03 12:46:58,978 Uncaught exception on thread Thread[ReadStage-2,5,node1]
java.lang.OutOfMemoryError: Java heap space
  at java.util.Arrays.copyOf(Arrays.java:3236)
  at java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:118)
  at java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:93)
  at java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:153)
  at java.io.PrintStream.write(PrintStream.java:480)
  at java.io.FilterOutputStream.write(FilterOutputStream.java:97)
  at ch.qos.logback.core.joran.spi.ConsoleTarget$1.write(ConsoleTarget.java:37)
  at ch.qos.logback.core.OutputStreamAppender.writeBytes(OutputStreamAppender.java:199)
  at ch.qos.logback.core.OutputStreamAppender.subAppend(OutputStreamAppender.java:231)
  at ch.qos.logback.core.OutputStreamAppender.append(OutputStreamAppender.java:102)
  at ch.qos.logback.core.UnsynchronizedAppenderBase.doAppend(UnsynchronizedAppenderBase.java:84)
  at ch.qos.logback.core.spi.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:51)
  at ch.qos.logback.classic.Logger.appendLoopOnAppenders(Logger.java:270)
  at ch.qos.logback.classic.Logger.callAppenders(Logger.java:257)
  at ch.qos.logback.classic.Logger.buildLoggingEventAndAppend(Logger.java:421)
  at ch.qos.logback.classic.Logger.filterAndLog_0_Or3Plus(Logger.java:383)
  at ch.qos.logback.classic.Logger.error(Logger.java:522)
  at org.apache.cassandra.service.StorageProxy$LocalReadRunnable.runMayThrow(StorageProxy.java:1893)
  at org.apache.cassandra.service.StorageProxy$DroppableRunnable.run(StorageProxy.java:2585)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:162)
  at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecutorService.java:134)
  at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:119)
  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  at java.lang.Thread.run(Thread.java:748)
```


Here is my diff to hit this

```
$ git diff
diff --git a/build.xml b/build.xml
index f8183ae320..4742105c70 100644
--- a/build.xml
+++ b/build.xml
@@ -219,10 +219,12 @@
         <string>-XX:MaxMetaspaceExpansion=64M</string>
         <string>-XX:MaxMetaspaceSize=512M</string>
         <string>-XX:MetaspaceSize=128M</string>
+        <string>-XX:-OmitStackTraceInFastThrow</string>
     </resources>
     <pathconvert property="_jvm8_test_arg_items_concat" refid="_jvm8_test_arg_items" pathsep=" "/>
     <resources id="_jvm11_test_arg_items">
         <string>-XX:-CMSClassUnloadingEnabled</string>
+        <string>-XX:-OmitStackTraceInFastThrow</string>
     </resources>
     <pathconvert property="_jvm11_test_arg_items_concat" refid="_jvm11_test_arg_items" pathsep=" "/>
     <condition property="test-jvmargs" value="${_jvm11_test_arg_items_concat}" else="${_jvm8_test_arg_items_concat}">
diff --git a/test/distributed/org/apache/cassandra/distributed/test/ReadFailureTest.java b/test/distributed/org/apache/cassandra/distributed/test/ReadFailureTest.java
index f3226c7d8b..f33b4e6709 100644
--- a/test/distributed/org/apache/cassandra/distributed/test/ReadFailureTest.java
+++ b/test/distributed/org/apache/cassandra/distributed/test/ReadFailureTest.java
@@ -92,7 +92,7 @@ public class ReadFailureTest extends TestBaseImpl
                 assertNotNull(String.format("Did not receive expected ReadFailureException. Instead caught %s\n%s",
                                             t, ExceptionUtils.getStackTrace(t)),
                               t.getMessage());
-                assertTrue(t.getMessage().contains(RequestFailureReason.READ_TOO_MANY_TOMBSTONES.name()));
+                assertTrue(t.getMessage(), t.getMessage().contains(RequestFailureReason.READ_TOO_MANY_TOMBSTONES.name()));
             }
         }
     }
```

Reply:
CI runs differently than IntelliJ and the jvm-dtest ant tasks, so to reproduce what I did here is my script (which mimics CI (jenkins and circle)).

```
#!/usr/bin/env bash

#set -o xtrace
set -o errexit
set -o pipefail
set -o nounset

usage() {
  if [[ $# -gt 0 ]]; then
    echo "$*" 1>&2
  fi
  cat <<EOF
Usage: $(basename $0) [class name] (options)*
Options:
--cdc           - runs with cdc enabled
--compression   - run with compression enabled
--reuse         - skip calling ant realclean
EOF
  exit 1
}

if [[ $# -eq 0 ]]; then
  usage "Missing required arguments"
fi

class_name="$1"
shift
path="$(echo "$class_name" | tr '.' '/' ).java"
prefix="$( find test | grep "$path" | awk -F/ '{print $2}' )"

task="testclasslist"
fresh_build=true
while [[ "$#" -gt 0 ]]; do
  case "$1" in
    --cdc)
      task="testclasslist-cdc"
      shift
      ;;
    --compression)
      task="testclasslist-compression"
      shift
      ;;
    --reuse)
      fresh_build=false
      shift
      ;;
    *)
      usage "Unknown argument $1"
      ;;
  esac
done

if [[ "$fresh_build" == true ]]; then
  if [[ ! "$path" =~ distributed/upgrade ]]; then
    ant realclean && ant && ant generate-idea-files
  fi
fi

# cleanup logs so w/e is around are for this run
rm -rf build/test/logs || true

test_timeout=$(grep "name=\"test.$prefix.timeout\"" build.xml | awk -F'"' '{print $4}' || true)
if [ -z "$test_timeout" ]; then
  test_timeout=$(grep 'name="test.timeout"' build.xml | awk -F'"' '{print $4}')
fi
ant "$task" -Dtest.timeout="$test_timeout" -Dtest.classlistfile=<(echo "$path") -Dtest.classlistprefix=$prefix

```

Reply:
I ran this in a loop while listening to music and never saw this fail

```
>>> Running attempt 84
```

Looks stable now (didn't lower resources)

Reply:
tested without the change and saw it is replicating, and with the change its stable.